hub/core/chunk.py

```diff
@@ -110,29 +193,65 @@ def register_sample_to_headers(
             ValueError: If `incoming_num_bytes` is not divisible by `num_samples`.
         """
 
-        num_bytes_per_sample = incoming_num_bytes
         self.shapes_encoder.register_samples(sample_shape, 1)
-        self.byte_positions_encoder.register_samples(num_bytes_per_sample, 1)
+        if (
+            incoming_num_bytes is not None
+        ):  # incoming_num_bytes is not applicable for non lz4 compressions
+            self.byte_positions_encoder.register_samples(incoming_num_bytes, 1)
+        self._clear_decompressed_cache()
 
     def update_sample(
-        self, local_sample_index: int, new_buffer: memoryview, new_shape: Tuple[int]
+        self,
+        local_sample_index: int,
+        new_buffer: memoryview,
+        new_shape: Tuple[int],
+        chunk_compression: Optional[str] = None,
+        dtype: Optional[np.dtype] = np.dtype("uint8"),
     ):
         """Updates data and headers for `local_sample_index` with the incoming `new_buffer` and `new_shape`."""
 
         expected_dimensionality = len(self.shapes_encoder[local_sample_index])
         if expected_dimensionality != len(new_shape):
             raise TensorInvalidSampleShapeError(new_shape, expected_dimensionality)
-
         new_nb = len(new_buffer)
+        self.shapes_encoder[local_sample_index] = new_shape
+        if chunk_compression:
+            if chunk_compression == "lz4":
+                decompressed_buffer = self.decompressed_data()
```

Comment:
yeah, we definitely need a better way to do this. please consider multiple chunk classes. also, i don't think it's good to have this if statement for chunk compressions. ideally this should be more modular

Reply:
mostly because backwards compatibility will be a huge pain with these if blocks everywhere